contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,866 @@
+        validatorExist(operatorAddress)
+        whenNotPaused
+    {
+        uint256 bnbAmount = msg.value;
+        require(bnbAmount >= minDelegationBNBChange, "INVALID_DELEGATION_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        address pool = valInfo.poolModule;
+        require(_checkPoolImplementation(pool), "IMPLEMENTATION_NOT_MATCH");
+
+        if (valInfo.jailed || IStakePool(pool).getSelfDelegationBNB() < minSelfDelegationBNB) {
+            // only self delegation
+            require(delegator == operatorAddress, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 shares = IStakePool(pool).delegate{ value: bnbAmount }(delegator);
+        emit Delegated(operatorAddress, delegator, shares, bnbAmount);
+
+        IGovBNB(govBNB).mint(operatorAddress, delegator, bnbAmount);
+    }
+
+    function undelegate(
+        address operatorAddress,
+        uint256 shares
+    ) public onlyInitialized validatorExist(operatorAddress) whenNotPaused {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        require(
+            IStakePool(valInfo.poolModule).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_UNDELEGATION_AMOUNT"
+        );
+
+        uint256 bnbAmount = IStakePool(valInfo.poolModule).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (
+            delegator == operatorAddress && IStakePool(valInfo.poolModule).getSelfDelegationBNB() < minSelfDelegationBNB
+        ) {
+            _validators[operatorAddress].jailed = true;
+            _removeEligibleValidator(valInfo.consensusAddress);
+            emit ValidatorJailed(operatorAddress);
+        }
+
+        IGovBNB(govBNB).burn(operatorAddress, delegator, bnbAmount);
+    }
+
+    function redelegate(
+        address srcValidator,
+        address dstValidator,
+        uint256 shares
+    )
+        public
+        onlyInitialized
+        validatorExist(srcValidator)
+        validatorExist(dstValidator)
+        validatorNotJailed(dstValidator)
+        whenNotPaused
+    {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = _validators[srcValidator];
+        Validator memory dstValInfo = _validators[dstValidator];
+        require(
+            IStakePool(srcValInfo.poolModule).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_REDELEGATION_AMOUNT"
+        );
+
+        uint256 bnbAmount = IStakePool(srcValInfo.poolModule).unbond(delegator, shares);
+        uint256 newShares = IStakePool(dstValInfo.poolModule).delegate{ value: bnbAmount }(delegator);
+        emit Redelegated(srcValidator, dstValidator, delegator, shares, newShares, bnbAmount);
+
```

Comment:
We need to check whether it is a self delegator from the srcvalidator, and whether it is under the minselfdelegation now.

Reply:
Done